crates/libm-test/src/gen/random.rs

```diff
@@ -0,0 +1,95 @@
+//! A simple generator that produces deterministic random input, caching to use the same
+//! inputs for all functions.
+
+use std::sync::LazyLock;
+
+use rand::{Rng, SeedableRng};
+use rand_chacha::ChaCha8Rng;
+
+use super::CachedInput;
+use crate::GenerateInput;
+
+const SEED: [u8; 32] = *b"3.141592653589793238462643383279";
+
+/// Number of tests to run.
+const NTESTS: usize = {
+    let mut ntests = if cfg!(optimizations_enabled) { 5000 } else { 500 };
+
+    // Tests seem to be pretty slow on non-64-bit targets, emulated ppc, and x86 MacOS
+    if !cfg!(target_pointer_width = "64")
+        || cfg!(target_arch = "powerpc64")
+        || cfg!(all(target_arch = "x86_64", target_vendor = "apple"))
+    {
+        ntests /= 5;
+    }
+
+    ntests
+};
+
+/// Tested inputs.
+static TEST_CASES: LazyLock<CachedInput> = LazyLock::new(|| make_test_cases(NTESTS));
+
+/// The first argument to `jn` and `jnf` is the number of iterations. Make this a reasonable
+/// value so tests don't run forever.
+static TEST_CASES_JN: LazyLock<CachedInput> = LazyLock::new(|| {
+    // It is easy to overflow the stack with these in debug mode
+    let iterations = if cfg!(optimizations_enabled) && cfg!(target_pointer_width = "64") {
+        0xffff
+    } else if cfg!(windows) {
+        0x00ff
+    } else {
+        0x0fff
+    };
+
+    let mut cases = (&*TEST_CASES).clone();
+    for case in cases.inputs_i32.iter_mut() {
+        case.0 = iterations;
+    }
+    for case in cases.inputs_i32.iter_mut() {
+        case.0 = iterations;
+    }
+    cases
+});
+
+fn make_test_cases(ntests: usize) -> CachedInput {
+    let mut rng = ChaCha8Rng::from_seed(SEED);
+
+    // make sure we include some basic cases
+    let mut inputs_i32 = vec![(0, 0, 0), (1, 1, 1), (-1, -1, -1)];
+    let mut inputs_f32 = vec![
+        (0.0, 0.0, 0.0),
+        (f32::EPSILON, f32::EPSILON, f32::EPSILON),
+        (f32::INFINITY, f32::INFINITY, f32::INFINITY),
+        (f32::NEG_INFINITY, f32::NEG_INFINITY, f32::NEG_INFINITY),
+        (f32::MAX, f32::MAX, f32::MAX),
+        (f32::MIN, f32::MIN, f32::MIN),
+        (f32::MIN_POSITIVE, f32::MIN_POSITIVE, f32::MIN_POSITIVE),
+        (f32::NAN, f32::NAN, f32::NAN),
+    ];
+    let mut inputs_f64 = vec![
+        (0.0, 0.0, 0.0),
+        (f64::EPSILON, f64::EPSILON, f64::EPSILON),
+        (f64::INFINITY, f64::INFINITY, f64::INFINITY),
+        (f64::NEG_INFINITY, f64::NEG_INFINITY, f64::NEG_INFINITY),
+        (f64::MAX, f64::MAX, f64::MAX),
+        (f64::MIN, f64::MIN, f64::MIN),
+        (f64::MIN_POSITIVE, f64::MIN_POSITIVE, f64::MIN_POSITIVE),
+        (f64::NAN, f64::NAN, f64::NAN),
+    ];
+
+    inputs_i32.extend((0..(ntests - inputs_i32.len())).map(|_| rng.gen::<(i32, i32, i32)>()));
+    inputs_f32.extend((0..(ntests - inputs_f32.len())).map(|_| rng.gen::<(f32, f32, f32)>()));
+    inputs_f64.extend((0..(ntests - inputs_f64.len())).map(|_| rng.gen::<(f64, f64, f64)>()));
```

Comment:
These are generating only floats in the range `0.0..1.0`, and IIRC not even all of those, for efficiency reasons. Generating integers and mapping with `from_bits` is the quick fix, though it's probably not the ideal distribution either. What would actually make sense depends on function quite a bit. As in, you want to give `sqrt` some negative inputs to test that it returns NaN, but benchmarking it with 50% negative inputs doesn't make much sense.

Reply:
Well that is a big oversight on random number generation, thanks for the catch.

This is just meant to be a generator that is quick and easy to review to get the infrastructure merged, even though it isn't a very good test. I am working on something that will be a bit smarter and take function domain, periodicity, inflection points, etc into account, but that will be a follow up https://github.com/tgross35/rust-libm/blob/e8ca082c19972c680f23fed2915d5ca19543f20f/crates/libm-test/src/domain.rs (that file is just the very basics).

It is still meant to work right though, I'll make the randomness change.

Reply:
One idea I've often used to try to generate a better ratio of interesting inputs is to pick the most important bits to vary exhaustively, and fill in the rest from just a few choices. Like for an integer you might generate all combinations of the 4 high bits and 4 low bits, and fill the middle bits from {all 0s, all 1s, a handful of random patterns}. For a float you also want the exponent low bits and mantissa high bits, so it get a bit messy.

Reply:
My complete idea is to combine the following:

* For each "interesting" point, check every value within ~500 ULP. This will catch behavior at zero, infinities, poles, inflection points, and domain boundaries.
* Do something like a logspace within the function's domain or period to get an even distribution for a certain number of iterations.
* For periodic functions, do some kind of check that they repeat (maybe? or check a couple periods toward the infinities).
* Try a number of random outputs that are not in the function's domain.

I'll also add exhaustive tests that check every value within the domain, behind a gate. Plus full exhaustive tests just to have them.

@beetrees was suggesting we could run exhaustive tests in CI if the relevant source file was changed.

Reply:
For reference, I am kind of basing the idea off of what I did for test-float-parse where there is a common interface, then you have a variety of different generators that can be plugged in https://github.com/rust-lang/rust/tree/master/src/etc/test-float-parse/src/gen

Reply:
I wrote a bit of code for the above idea, and with another look it ended up neater than I expected, so maybe this can come in handy:
```rust
/// Iterator of all bitpatterns that are a subset of the mask.
fn bit_subsets(mask: u32) -> impl Iterator<Item = u32> {
    let ignore = !mask;
    assert!(ignore != 0, "mask can't have all bits set");
    let mut counter = ignore.wrapping_sub(1);
    std::iter::from_fn(move || {
        // Incrementing carries through any ignored bits, which
        // are then toggled back on
        counter = counter.checked_add(1)? | ignore;
        Some(counter & !ignore)
    })
}
```

Reply:
Well, fixing the random number generation produced quite a few more failures. It will be more interesting for the MPFR PR, seems like musl has some unusual edge cases here.